Adopting CNCF-style documentation for our org structure (#15)
k8sgateway/community: CONTRIBUTOR_LADDER.md

* Inactivity is measured by:
    * Periods of no contributions for longer than 12 months
    * Periods of no communication for longer than 6 months
* Consequences of being inactive include:
    * Involuntary removal or demotion
    * Being asked to move to Emeritus status

## Involuntary Removal or Demotion

Involuntary removal/demotion of a contributor happens when responsibilities and requirements aren't being met. This may include repeated patterns of inactivity, extended period of inactivity, a period of failing to meet the requirements of your role, abusing the CI/CD priviedges and/or a violation of the Code of Conduct. This process is important because it protects the community and its deliverables while also opens up opportunities for new contributors to step in.

Involuntary removal or demotion is handled through a vote by a majority of the current Maintainers.

## Stepping Down/Emeritus Process

If and when contributors' commitment levels change, contributors can consider stepping down (moving down the contributor ladder) vs moving to emeritus status (completely stepping away from the project).

> Contact the Maintainers about changing to Emeritus status, or reducing your contributor level.

Reviewer: Should we be more explicit, and say to do this on the CNCF slack in a specific channel?

Author: Probably good. We should consider updating the contributor docs to have a section about where to contact us on Slack which we can link to easily. Right now the slack information is in an ordered list and linking to it would be tricky.

Reviewer 2: Hi Kevin, thanks for the review. I've added CNCF slack here.

# K8sGateway Contributor Ladder

This document describes the set of roles individuals may have within the community, the requirements of each role, and the contributor ladder for someone to grow.

* [Contributor Ladder](#contributor-ladder)
    * [Contributor](#contributor)
    * [Organization Member](#organization-member)
    * [Reviewer](#reviewer)
    * [Maintainer](#maintainer)

## Contributor Ladder

> Hello! We are excited that you want to learn more about our project contributor ladder! This contributor ladder outlines the different contributor roles within the project, along with the responsibilities and privileges that come with them. Community members generally start at the first levels of the "ladder" and advance up it as their involvement in the project grows.  Our project members are happy to help you advance along the contributor ladder.

Reviewer 3: ```suggestion
This contributor ladder outlines the different contributor roles within the project, along with the responsibilities and privileges that come with them. Community members generally start at the first levels of the "ladder" and advance up it as their involvement in the project grows.  Our project members are happy to help you advance along the contributor ladder.
```

less wow!

Each of the contributor roles below is organized into lists of three types of things. "Responsibilities" are things that a contributor is expected to do. "Requirements" are qualifications a person needs to meet to be in that role, and "Privileges" are things contributors on that level are entitled to.

### Contributor

A Contributor contributes directly to the project and adds value to it. Contributions need not be code. People at the Contributor level may be new contributors, or they may only contribute occasionally.

* Responsibilities include:
    * Follow the [CNCF CoC](https://github.com/cncf/foundation/blob/main/code-of-conduct.md)

>     * Follow the project [contributing guide](CONTRIBUTING.md)

Reviewer 3: nit: "following"

Reviewer 2: resolved

* Requirements (one or several of the below):
    * Report issues
    * Show up at meetings, takes notes
    * Answer questions from other community members
    * Submit feedback on issues and PRs
    * Test releases and patches and submit reviews
    * Run or help run events
    * Promote the project in public

### Organization Member

An Organization Member is an established contributor who regularly participates in the project. Organization Members have privileges in the project repositories, and as such are expected to act in the interests of the whole project.

An Organization Member must meet the responsibilities and has the requirements of a Contributor, plus:

* Responsibilities include:
    * Continues to contribute regularly, as demonstrated by having at least 3 GitHub contributions per year, as shown by Devstats.
* Requirements:
    * Must have successful contributions to the project, including at least one merged PR.

>     * Member of the #k8sgateway channel on the [CNCF slack](https://slack.cncf.io).

Reviewer 3: "Must be a member"
```suggestion
    * Must be amember of the #k8sgateway channel on the [CNCF Slack](https://slack.cncf.io).
```

>     * Must have enabled 2FA on the GitHub account.

Reviewer 3: ```suggestion
    * Must have 2FA enabled on your GitHub account.
```

* Privileges:
    * May be assigned Issues and Reviews.
    * May give commands to CI/CD automation.
    * May label/triage issues.

> The process for a Contributor to become an Organization Member is as follows:
> 

Reviewer 3: ```suggestion
```

> If you'd like to become a member of the k8sgateway organization on GitHub, please fill out this form and submit it as a GitHub issue in the community repo. Give us a few days to review and you should receive an invitation to join.

Reviewer 3: ```suggestion
To apply to become a member of the k8sgateway organization on GitHub, please fill out this form and submit it as a GitHub issue in the community repo. Give us a few days to review and you should receive an invitation to join.
```

Who is us?  Who can accept this?  One maintainer?
Why is this form in this doc?  Can we just link to a GH boilerplate issue?

Reviewer 2: Done :)

GitHub user id:

Company affiliation:

Requirements:

- [ ] I have reviewed the project [contributing guide](CONTRIBUTING.md)
- [ ] I have enabled 2FA on my GitHub account
- [ ] I have joined #k8sgateway channel on the [CNCF slack](https://slack.cncf.io)

Provide a link to at least one PR that you have successfully pushed to one
of the repos:

----------------------

### Reviewer

A Reviewer has responsibility for specific code, documentation, test, or other project areas. They are collectively responsible, with other Reviewers, for reviewing all changes to those areas and indicating whether those changes are ready to merge. They have a track record of contribution and review in the project.

Reviewers are responsible for a "specific area." This can be a specific code directory, chapter of the docs, test job, event, or other clearly-defined project component that is smaller than an entire repository or subproject. Most often it is one or a set of directories in one or more Git repositories. The "specific area" below refers to this area of responsibility.

Reviewers have all the rights and responsibilities of an Organization Member, plus:

* Responsibilities include:

>     * Reviewing at least 3 PRs per year

Reviewer 3: that's comically low if you have to have reviewed 20 just to get in

Reviewer 2: Yes it is a bit low, but I don't want to set to too high.  In Istio, if you are still around with reviewing a few PRs per year, you get to keep your maintainer status.

    * Helping other contributors become reviewers
* Requirements:
    * Experience as a Contributor for at least 3 months

>     * Has reviewed, or helped review, at least 20 Pull Requests

Reviewer 3: you must... be a reviewer before you can become a reviewer?  🤔

Author: I think its a case of any org member can review (leave comments, click the approve button) but until you are officially a reviewer said approval doesn't unblock merging.

Reviewer 2: yep, you have to review a good amount of PRs before you can become an official reviewer.

    * Has demonstrated an in-depth knowledge of the specific area
* Additional privileges:
    * Has GitHub or CI/CD rights to approve pull requests in specific directories
    * Can recommend and review other contributors to become Reviewers

The process of becoming a Reviewer is:
1. The contributor is nominated by opening a PR against the appropriate repository, which adds their GitHub username to the OWNERS file for one or more directories.

> 2. At least two members of the team that owns that repository or main directory, who are already Approvers, approve the PR.

Reviewer 3: We don't have an Approver role

Reviewer 2: i mean maintainers, updated

> ### Maintainer

Reviewer 3: I don't think that this piece of CNCF boilerplate is very well thought out.  For starters, I find it strange the path to what is effectively "being on the board of directors of the project" is through code review?

Reviewer 4: +1... the requirement should be related to supporting the project from a business standpoint, i.e., providing a distro.

Author: I'm not sure I understand this thread. Is the issue effectively that we don't want a maintainer council style of governance or that we don't have some other path to maintainer?

Reviewer 2: Providing builds is a requirement, which we will address in a separate PR.  It is not a required individual maintainer responsibility for someone to reach the maintainer ladder.

Re Craig's point, I think we can start with maintainers are the ones who doing the work and decide on strategy and priorities of the project, which is common for many OSS projects out there.

Reviewer 2: Opened https://github.com/k8sgateway/community/issues/27 for tracking build separately.

Maintainers are very established contributors who are responsible for the entire project. As such, they have the ability to approve PRs against any area of the repository or repositories they maintain, and are expected to participate in making decisions about the strategy and priorities of the project.

A Maintainer must meet the responsibilities and requirements of a Reviewer, plus:

* Responsibilities include:
    * Reviewing at least 3 PRs per year, especially PRs that involve multiple parts of the project
    * Mentoring new Reviewers
    * Determining strategy and policy for the project
* Requirements
    * Experience as a Reviewer for at least 3 months
    * Have at least 20 substantial PRs merged to the codebase.
    * Demonstrates a broad knowledge of the project across multiple areas
* Additional privileges:
    * Represent the project in public as a Maintainer
    * Have a vote in Maintainer decision-making discussions
    
Process of becoming a maintainer:

> 1. Any current Maintainer may nominate a current Reviewer to become a new Maintainer, by opening a PR against the root of the [TODO: main repository name] adding the nominee as an Approver in the OWNERS file.

Reviewer 3: [TODO: main repository name]

Reviewer 2: clarified it is the community repo.

> # K8sGateway Contributor Ladder

Reviewer 4: Will this file be xref'd from https://github.com/k8sgateway/k8sgateway/blob/main/CONTRIBUTING.md?

Reviewer 2: Yes good idea!

Each of the contributor roles below is organized into lists of three types of things. "Responsibilities" are things that a contributor is expected to do. "Requirements" are qualifications a person needs to meet to be in that role, and "Privileges" are things contributors on that level are entitled to.

### Contributor

A Contributor contributes directly to the project and adds value to it. Contributions need not be code. People at the Contributor level may be new contributors, or they may only contribute occasionally.

* Responsibilities include:
    * Follow the [CNCF CoC](https://github.com/cncf/foundation/blob/main/code-of-conduct.md)
    * Follow the project [contributing guide](CONTRIBUTING.md)
* Requirements (one or several of the below):
    * Report issues
    * Show up at meetings, takes notes
    * Answer questions from other community members
    * Submit feedback on issues and PRs
    * Test releases and patches and submit reviews

>     * Run or helps run events

Reviewer 4: nit: s/helps/help/

>     * Run or helps run events

Reviewer 2: ```suggestion
    * Run or help run events
```

    * Promote the project in public

### Organization Member

An Organization Member is an established contributor who regularly participates in the project. Organization Members have privileges in the project repositories, and as such are expected to act in the interests of the whole project.

An Organization Member must meet the responsibilities and has the requirements of a Contributor, plus:

* Responsibilities include:
    * Continues to contribute regularly, as demonstrated by having at least 3 GitHub contributions per year, as shown by Devstats.
* Requirements:

>     * Must have successful contributions to the project, including at least one of the following:

Author: 2 factor for your github account

Author: It's mentioned below but not here, maybe that's OK but seems off

Reviewer 4: Should add'l context be provided for a contribution, i.e. 3 PRs/yr to fix a typo IMHO is not enough.

Reviewer 2: Added 2 factor.  Understood 3 PRs per year for a typo isn't great, but would like to keep it simple for now and hope people won't abuse the system to just sending typos.  We can add more clarifications if someone abuses the system.

        * 1 merged PR.
    * Member of the #k8sgateway channel on the [CNCF slack](https://slack.cncf.io).
* Privileges:
    * May be assigned Issues and Reviews

>     * May give commands to CI/CD automation

Reviewer 2: added

The process for a Contributor to become an Organization Member is as follows:

> If you'd like to become a member of the Istio organization on GitHub, please fill out this form and submit it as a GitHub issue in the community repo. Give us a few days to review and you should receive an invitation to join.

Reviewer 5: ```suggestion
If you'd like to become a member of the k8sgateway organization on GitHub, please fill out this form and submit it as a GitHub issue in the community repo. Give us a few days to review and you should receive an invitation to join.
```

Reviewer 4: Consider removing "a few days".

> If you'd like to become a member of the Istio organization on GitHub, please fill out this form and submit it as a GitHub issue in the community repo. Give us a few days to review and you should receive an invitation to join.

Reviewer 5: can we link to an issue template? as an example of how it's done in the [k8s org](https://github.com/kubernetes/community/blob/master/community-membership.md#requirements), you open your own PR using a template requesting to become an org member, and list 2 sponsors (which need to be from different companies) , and those people need to sign off on the PR. do we want to do something like that, either for org membership and/or maintainer/other status?

Reviewer 2: I don't think we should follow k8s org because it is complex and has thousands of contributors. I prefer us follow a simple way which is one merged PR and open a request for applying membership.

I will open an issue template for this.

Reviewer 2: https://github.com/k8sgateway/community/pull/15/files good suggestion

----------------------
GitHub user id:

Company affiliation:

Requirements:

> - [ ] I have reviewed the the [CNCF CoC](https://github.com/cncf/foundation/blob/main/code-of-conduct.md)

Author: reviewed and agree to abide by/follow the CoC?

Reviewer 2: Longer term, i don't think we even need this because a merged PR will be guarded by review of CoC.

> - [ ] I have reviewed the project [contributing guide](CONTRIBUTING.md)

Reviewer 5: should they also sign a CLA before contributing code? similar to steps outlined here https://github.com/kubernetes/community/blob/master/CLA.md

Author: Good point

Reviewer 2: Yes, they would, so I will remove the CLA checklist given a merged PR is required.

- [ ] I have enabled 2FA on my GitHub account
- [ ] I have joined #k8sgateway channel on the [CNCF slack](https://slack.cncf.io)

Provide a link to at least one PR that you have successfully pushed to one
of the repos:

----------------------

### Reviewer

A Reviewer has responsibility for specific code, documentation, test, or other project areas. They are collectively responsible, with other Reviewers, for reviewing all changes to those areas and indicating whether those changes are ready to merge. They have a track record of contribution and review in the project.

> Reviewers are responsible for a "specific area." This can be a specific code directory, driver, chapter of the docs, test job, event, or other clearly-defined project component that is smaller than an entire repository or subproject. Most often it is one or a set of directories in one or more Git repositories. The "specific area" below refers to this area of responsibility.

Author: ```suggestion
Reviewers are responsible for a "specific area." This can be a specific code directory, chapter of the docs, test job, event, or other clearly-defined project component that is smaller than an entire repository or subproject. Most often it is one or a set of directories in one or more Git repositories. The "specific area" below refers to this area of responsibility.
```

Reviewers have all the rights and responsibilities of an Organization Member, plus:

* Responsibilities include:
    * Reviewing at least 3 PRs per year
    * Helping other contributors become reviewers
* Requirements:
    * Experience as a Contributor for at least 3 months

>     * Is an Organization Member

Author: Is being an org member part and parcel of being a contributor so it doesn't need its own line item. Maybe we just add a blurb about meets all requirements for lower rungs if we want to ensure that is clear rather than implied.

Reviewer 4: >Maybe we just add a blurb about meets all requirements for lower rungs

+1

    * Has reviewed, or helped review, at least 20 Pull Requests
    * Has demonstrated an in-depth knowledge of the specific area
* Additional privileges:
    * Has GitHub or CI/CD rights to approve pull requests in specific directories
    * Can recommend and review other contributors to become Reviewers

The process of becoming a Reviewer is:

> 1. The contributor is nominated by opening a PR against the appropriate repository, which adds their GitHub username to the OWNERS file for one or more directories.

Reviewer 2: That is a good point. This made me think we should remove reviewer as we don't have process in place for it.

> 1. The contributor is nominated by opening a PR against the appropriate repository, which adds their GitHub username to the OWNERS file for one or more directories.

Reviewer 4: [xref](https://github.com/k8sgateway/community/issues/24#issuecomment-2523655179) regarding owners file management.

> 1. The contributor is nominated by opening a PR against the appropriate repository, which adds their GitHub username to the OWNERS file for one or more directories.

Reviewer 2: Thanks! will resolve this - and revisit this when the https://github.com/k8sgateway/community/issues/24 is fixed.

2. At least two members of the team that owns that repository or main directory, who are already Approvers, approve the PR.

### Maintainer

> Maintainers are very established contributors who are responsible for the entire project. As such, they have the ability to approve PRs against any area of the project, and are expected to participate in making decisions about the strategy and priorities of the project.

Author: ```suggestion
Maintainers are very established contributors who are responsible for the entire project. As such, they have the ability to approve PRs against any area of the repository or repositories they maintain, and are expected to participate in making decisions about the strategy and priorities of the project.
```

Reviewer 4: > by opening a PR against the appropriate repository

Instead of contributors nominating themselves, should we consider that the nomination comes from a maintainer?

Author: I think it would be nice to get consistency between reviewer process and maintainer process. I'm not against a self-nomination so it's OK with me either way we land.

Along those lines, maintainer nominate maintainers so would we allow a reviewer (or maintainer of course) to nominate another reviewer or always a maintainer. Again, not really advocating either, just wanting to be a clear and consistent as we can be.

A Maintainer must meet the responsibilities and requirements of a Reviewer, plus:

* Responsibilities include:
    * Reviewing at least 3 PRs per year, especially PRs that involve multiple parts of the project
    * Mentoring new Reviewers
    * Determining strategy and policy for the project
* Requirements
    * Experience as a Reviewer for at least 3 months
    * Have at least 20 substantial PRs merged to the codebase.
    * Demonstrates a broad knowledge of the project across multiple areas
* Additional privileges:
    * Represent the project in public as a Maintainer

>     * Have a vote in Maintainer decision-making meetings

Author: Should we call this "maintainer council" or something like that?

Reviewer 2: I'm good as it is as this is from the template.

    
Process of becoming a maintainer:

> 1. Any current Maintainer may nominate a current Reviewer to become a new Maintainer, by opening a PR against the root of the [TODO: main repository name] adding the nominee as an Approver in the OWNERS file.

Author: Is it OK for a reviewer to self-nominate? Should they ask in a community meeting and then a maintainer will "sponsor" them by opening the PR?

Reviewer 4: > Is it OK for a reviewer to self-nominate?

-1

Reviewer 2: Agreed, needs to be nominated by a maintainer.

> 1. Any current Maintainer may nominate a current Reviewer to become a new Maintainer, by opening a PR against the root of the [TODO: main repository name] adding the nominee as an Approver in the OWNERS file.

Author: 1. Does the maintainer making the nomination need to be of the speculation the nomination is for? Concretely, would a community maintainer nominating someone for controller maintainer be allowed?
2. Same question for approving. If a community maintainer approves a request to make someone a controller maintainer is that a "valid" approval?

It seems to me the answer for question 1 is maybe. For the second question I would think the answer would be a no and the approvals should come from folks who are active maintainers of the specialization. (Probably also objections?)

Reviewer 2: clarified... agree 2 should be maintainer for that part of the project.

<!-- template begins here -->

* [Contributor Ladder](#contributor-ladder)
    * [Contributor](#contributor)
    * [Organization Member](#organization-member)
    * [Reviewer](#reviewer)
    * [Maintainer](#maintainer)
* [Inactivity](#inactivity)
* [Involuntary Removal](#involuntary-removal-or-demotion)
* [Stepping Down/Emeritus Process](#stepping-downemeritus-process)
* [Contact](#contact)

## Contributor Ladder

Hello! We are excited that you want to learn more about our project contributor ladder! This contributor ladder outlines the different contributor roles within the project, along with the responsibilities and privileges that come with them. Community members generally start at the first levels of the "ladder" and advance up it as their involvement in the project grows.  Our project members are happy to help you advance along the contributor ladder.

Each of the contributor roles below is organized into lists of three types of things. "Responsibilities" are things that a contributor is expected to do. "Requirements" are qualifications a person needs to meet to be in that role, and "Privileges" are things contributors on that level are entitled to.

> ### Contributor

Author: It probably makes sense to have an entry point which does not automatically grant having CI run against your PR and has no capability to interact with CI just to save on scarce resources.

Reviewer 2: We discussed this in the community mtg this week. This is not something we'd worry atm. If we see someone abuse the system, we will remove them.

<!--TODO: project leads to fill in exact details of this role for your project-->
Description: An Organization Member is an established contributor who regularly participates in the project. Organization Members have privileges in both project repositories and elections, and as such are expected to act in the interests of the whole project.

An Organization Member must meet the responsibilities and has the requirements of a Contributor, plus:

* Responsibilities include:
    * Continues to contribute regularly, as demonstrated by having at least [TODO: Number] [TODO: Metric] a year, as demonstrated by [TODO: contributor metrics source]. <!-- Example: "as demonstrated by having at least 50 GitHub contributions per year, as shown by Devstats"-->
* Requirements:
    * Must have successful contributions to the project, including at least one of the following:
        * [TODO: Number] accepted PRs,
        * Reviewed [TODO: Number] PRs,
        * Resolved and closed [TODO: Number] Issues,
        * Become responsible for a key project management area,
        * Or some equivalent combination or contribution

>     * Must have been contributing for at least [TODO: Number] months

Author: This role will likely be able to use scarce project resources and so the level of trust with the project likely needs to be a little higher.

I am wondering about setting the threshold to 5-10 PRs/Reviews/Issue and 3-6 months of activity as a contributor?

Reviewer 2: resolving this as per our discussion if folks abuse the CI priviledge, we will remove them.

    * Must be actively contributing to at least one project area
    * Must have two sponsors who are also Organization Members, at least one of whom does not work for the same employer
    * [TODO: other requirements] <!--TODO: such as repository access or enabling 2FA on their GitHub account -->

* Privileges:
    * May be assigned Issues and Reviews
    * May give commands to CI/CD automation
    * Entitled to vote in the [TODO: appropriate election]
    * Can be added to [TODO: Repo Host] teams
    * Can recommend other contributors to become Org Members
    * [TODO: Other Privileges]

The process for a Contributor to become an Organization Member is as follows:

<!-- the process of becoming an organization member is going to depend strongly on how your project manages its infrastructure. TODO: Project leads to fill in exact details of how a contributor becomes an organization member-->
1.
2.

> 3.

Author: Someone (user? existing maintainer?) files a PR against /some/ file. 

1. a single approver?
2. simple majority of maintainers in the area?

> ### Reviewer

Reviewer 2: I'm okay to keep it, it does show a ladder of progression.

* Patch Maintainer: sometimes different from the release manager, owns the tooling, team, and schedule for patching stable releases. [Example](https://github.com/kubernetes/sig-release/blob/master/release-engineering/role-handbooks/patch-release-team.md)
* Community Manager: owns social media, community metrics, new contributor process, and similar areas.
-->

## Inactivity
<!--TODO: project leads to fill in exact details for how you measure inactivity for your project-->
It is important for contributors to be and stay active to set an example and show commitment to the project. Inactivity is harmful to the project as it may lead to unexpected delays, contributor attrition, and a lost of trust in the project.

* Inactivity is measured by:
    * Periods of no contributions for longer than [TODO: Number] months
    * Periods of no communication for longer than [TODO: Number] months
* Consequences of being inactive include:
    * Involuntary removal or demotion

>     * Being asked to move to Emeritus status

Author: Pretty generous? 1 year?

<!-- TODO: replace with your method of removing/demoting contributors.  If you have a formal governance structure, this would be a good place to assign this to your governance, such as a Steering Committee.
Again, the example below is for a project without formal governance except the maintainers.-->
Involuntary removal or demotion is handled through a vote by a majority of the current Maintainers.

## Stepping Down/Emeritus Process
If and when contributors' commitment levels change, contributors can consider stepping down (moving down the contributor ladder) vs moving to emeritus status (completely stepping away from the project).

Contact the Maintainers about changing to Emeritus status, or reducing your contributor level.

## Contact
* For inquiries, please reach out to:

>     *  <!-- TODO: fill in contact info for appropriate group or person for contributor mentorship-->

Author: Slack channel I suppose. This might be a good reason to have a contributor slack channel.
